src/protocols/mod.rs

```diff
@@ -13,8 +14,169 @@
 
     @license GPL-3.0+ <https://github.com/KZen-networks/multi-party-ed25519/blob/master/LICENSE>
 */
+use curv::arithmetic::Converter;
+use curv::cryptographic_primitives::proofs::ProofError;
+use curv::elliptic::curves::{Ed25519, Point, Scalar};
+use curv::BigInt;
+use rand::{thread_rng, Rng};
+use sha2::{Digest, Sha512};
+
 // simple ed25519 based on rfc8032
 // reference implementation: https://ed25519.cr.yp.to/python/ed25519.py
 pub mod aggsig;
 pub mod multisig;
 pub mod thresholdsig;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ExpendedPrivateKey {
+    pub prefix: Scalar<Ed25519>,
+    private_key: Scalar<Ed25519>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ExpendedKeyPair {
+    pub public_key: Point<Ed25519>,
+    expended_private_key: ExpendedPrivateKey,
+}
+
+impl ExpendedKeyPair {
+    pub fn create() -> ExpendedKeyPair {
+        let secret = thread_rng().gen();
+        Self::create_from_private_key(secret)
+    }
+
+    pub fn create_from_private_key(secret: [u8; 32]) -> ExpendedKeyPair {
+        let h = Sha512::new().chain(secret).finalize();
+        let mut private_key: [u8; 32] = [0u8; 32];
+        let mut prefix: [u8; 32] = [0u8; 32];
+        prefix.copy_from_slice(&h[32..64]);
+        private_key.copy_from_slice(&h[0..32]);
+        private_key[0] &= 248;
+        private_key[31] &= 63;
```

Comment:
This part of bit clamping is different from what is implemented in the libsodium (see [this line](https://github.com/jedisct1/libsodium/blob/6d566070b48efd2fa099bbe9822914455150aba9/src/libsodium/crypto_sign/ed25519/ref10/keypair.c#L20)). Are they both correct?

Reply:
Yes. In both implementation you unset the top-most bit of the last octet and set the 2nd-top-most bit of the last octet.

The difference is that we first unset both two-top-most bits of the last octet and then set the 2nd-top-most bit of the last octet while in libsodium they unset the top-most-bit and set the 2nd-top-most bit.
I agree we should consider changing it from 63 to 127 just for the sake of following the RFC verbatim, but the resulting output seems correct to me.